### PR TITLE
Add support for smart string/copy paste when the destination is a raw string literal.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This includes properly fixing up the delimiters (adding `$` and `"` as necessayr) as well as handling indentation and also escaping internal braces on interpolations as necessary.

Lots of tests added.   Several hundred more to go there.

### CODE REVIEW

File: src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs [hunk 1]
```diff
@@ -83,24 +84,7 @@ private ImmutableArray<TextChange> GetEditsForNonRawString()
...
-            {
-                AppendContentForInterpolatedString(builder);
-            }
-            else
-            {
-                throw ExceptionUtilities.Unreachable;
-            }
-
-            return ImmutableArray.Create(new TextChange(_selectionBeforePaste.Span.ToTextSpan(), builder.ToString()));
-        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs:100]: this method called out to two methods that were almost identical in how they behaved.  so i merged those two methods into one, and then tweaked it in the one place they differ in behavior.

File: src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs [hunk 2]
```diff
@@ -132,38 +127,169 @@ private void AppendContentForLiteral(StringBuilder builder)
...
+            // the interpolation to affect the delimiters.  e.g. a interpolation containing `""""` *inside* of it
+            // doesn't require updating the delimiters of the outer expression.
+            //
+            // Also, after the text is inserted, look to see if we need to convert a single-line raw expression to
+            // multi-line.
+            //
+            // At this point, we will now have the information necessary to actually insert the content and do things 
+            // like give interpolations the proper number of braces for the final string we're making.
+
+            var dummyContentEdit = GetContentEditForRawString(insertInterpolations: false, dollarSignCount: -1, indentationWhitespace: "");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs:154]: this method is very long, but is described by the above comment.  i intend to break it apart into pieces to make it clearer as to what happens.

File: src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs [hunk 3]
```diff
@@ -132,38 +127,169 @@ private void AppendContentForLiteral(StringBuilder builder)
...
+                    if (!beforeSelection && interpolation.Span.Start < _selectionSpanBeforePaste.End)
+                        continue;
+
+                    edits.Add(new TextChange(new TextSpan(interpolation.OpenBraceToken.Span.End, 0), new string('{', dollarSignsToAdd)));
+                    edits.Add(new TextChange(new TextSpan(interpolation.CloseBraceToken.Span.Start, 0), new string('}', dollarSignsToAdd)));
+                }
+            }
+        }
+
+        private TextChange GetContentEditForRawString(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs:239]: ignore the diff in this mehtod.  it comes from the fact that i merged two similar methods into one (talked above above), but then added a method for the raw case.  the diff thinkgs the raw-method shoudl diff against one of the prior methods.

File: src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs [hunk 4]
```diff
@@ -132,38 +127,231 @@ private void AppendContentForLiteral(StringBuilder builder)
...
+            //
+            // First, just take the copied text and determine the most basic edit that would insert it into the
+            // destination string.  Importantly, do not insert interpolations in this step (instead just replace them
+            // with a dummy character).
+            //
+            // Second, after this text is inserted, look at the content regions of the string after the paste and look
+            // at the sequences of `"` and `{` in them to see if we need to update the delimiters of the raw string.
+            // Note: this is why it is critical that any interpolations are not inserted.  We don't want the content of
+            // the interpolation to affect the delimiters.  e.g. a interpolation containing `""""` *inside* of it
+            // doesn't require updating the delimiters of the outer expression.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs:146]:
```suggestion
            // doesn't require updating the delimiters of the outer raw-string expression.
```

**Comment by CyrusNajmabadi** [comment]
@davidwengier let me know what you think, and if there are ways to make things clearer.

**Review by davidwengier — APPROVED** [review]
Logic makes sense, though one thing that isn't clear (and I haven't looked through the tests) but is there coverage for the scenario where the newly pasted interpolations need to have braces added? eg, if pasting from a `$"""` to a `$$$"""` or something.

Though I imagine this stuff gets extremely complicated so its probably fair enough if some things just aren't supported.

**Comment by CyrusNajmabadi** [comment]
>  but is there coverage for the scenario where the newly pasted interpolations need to have braces added? eg, if pasting from a $""" to a $$$""" or something.

Yup.  Test for that *and* tests for ensuring that if we need to change the number of '$' that we also update *existing* interpolations.

**Pull request merged**